File: Kakao/2020공채/buildGidong.js
let N;
let map = [[]];

function getIndex( index){
    return N-index;
}

function canBuildGi(x,y){

    if(x<0 || x>N || y>0 || y>N) return false;
    //기둥위나 보의 한쪽끝이면
    if(map[x][y] > 0){
        return true;
    }

    //다른 보의 끝
    if(y-1 >= 0){
        if(map[x][y-1] == 2)
            return true;
    }

    return false;
}


function canBuildBo(x,y){

    //보가 바닥이면
    if(x == N) return false;

    if(x<0 || x>N || y>0 || y>N) return false;

    if(map[x+1][y] === 1 || map[x+1][y] === 3) return true;
    
    if(y == 0 || y == N) return false;

    if(map[x][y-1] >=2 && map[x][y+1] >=2) return true;

    return false;

}

function canDeleteGi(x,y){
    
    map[x][y] -= 1;
    
    if(map[x-1][y] ==1 || map[x-1][y] == 3)
        if(!canBuildGi(x-1,y)){
            map[x][y] +=1
            return false;
        }
    
    if(map[x-1][y] >=2){
        if(!canBuildBo(x-1,y)){
            map[x-1][y] +=1
            return false;
        }
    }
    
    if(map[x-2][y] >=2){
        if(!canBuildBo(x-1,y)){
            map[x-2][y] += 1;
            return false;
        }
    }

    return true;
    
}


function canDeleteBo(x,y){
    map[x][y] -= 2;
    if(map[x][y] ==1 || map[x][y] == 3){
        if(!canBuildGi(x,y)){
            map[x][y] += 2;
            return false;
        }
    }
    if(map[x][y+1] ==1 || map[x][y+1] == 3){
        if(!canBuildGi(x,y+1)){
            map[x][y] += 2;
            return false;
        }
    }

    if(map[x][y-1] >=2){
        if(!canBuildBo(x,y-1)){
            map[x][y-1]+= 2;
            return false;
        }
    }
    if(map[x][y+1] >=2){
        if(!canBuildBo(x,y+1)){
            map[x][y+1]+= 2;
            return false;
        }
    }

    

}

function solution(n, build_frame) {
    N = n;
    var answer = [[]];
    
    

    for(let i =0;i<n;i++){
        let temp  =[]
        for(let j =0;j<=n;j++){
            temp.push(0);
        }
        map.push(temp);
    }
    let temp = []
    for(let i =0;i<=n;i++){
        temp.push(2);
    }
    map.push(temp);

    for(let i =0;i<build_frame.length;i++){
        let x = getIndex(build_frame[i][0]);
        let y = build_frame[i][1];
        let a = build_frame[i][2];
        let b = build_frame[i][3];
        //구조물 설치
        if(b ===0){
            //기둥설치
            if(a == 0){
                if(canBuildGi(x,y)){
                    map[x][y] += 1;
                }
            }

            //보설치
            else if(a == 1){
                if(canBuildBo(x,y)){
                    map[x][y] +=2;
                }
            }

        }
        //구조물 삭제
        else{
            //기둥삭제
            if(a == 0){
                canDeleteGi(x,y);

            }

            //보 삭제
            else if(a == 1){
                canDeleteBo(x,y);
            }


        }
    }

    
    

    return map;
}

File: Kakao/2020공채/buildGidong.cpp
#include <string>
#include <vector>
#include<iostream>

using namespace std;

bool board[101][101][2];
int n;

bool buildGi(int y,int x){

    //기둥을 벗어나므로
    if(y == n) return false;

    //바닥위
    if(y == 0) return true;

    //다른 기둥 위
    if(board[y-1][x][0]) return true;

    //지금 아래 보위
    if(board[y][x][1]) return true;

    //다른 보 위
    if(x == 0) return false;
    if(board[y][x-1][1]) return true;
    

    return false;
}

bool buildBo(int y,int x){

    if(y == 0) return false;
    
    //지금 밑에 기둥
    if(board[y-1][x][0]){
        //cout<<"1"<<endl; 
        return true;
    }

    //다른 한쪽이 기둥
    if(x+1<=n){
        if(board[y-1][x+1][0]){
            //cout<<"2"<<endl; 
            return true;
        }
    }

    //양쪽 보사이
    if(x-1>=0 && x+1<=n){
        if(board[y][x-1][1] && board[y][x+1][1]){
            //cout<<"3"<<endl;  
            return true;
        }
    }

    return false;
}


void canRemove(int y,int x, int type){
    board[y][x][type] =false;
    //cout<<"type : " << type+1<<endl;

    for(int i =0;i<=n;i++){
        for(int j =0;j<=n;j++){
            
            if(board[i][j][0]){
             
                if(!buildGi(i,j)){
                
                   //cout<<i<<j<<"여기"<<endl;
                    board[y][x][type]= true;
                    return;
                }
            }
            if(board[i][j][1]){
                if(!buildBo(i,j)){
                    //ccout<<i<<j<<"여기"<<endl;
                    board[y][x][type] = true;;
                    return;
                }
            }
            
        
        }
    }

}

vector<vector<int>> solution(int N, vector<vector<int>> build_frame) {
    n = N;
    vector<vector<int>> answer;

    for(int i =0;i<build_frame.size();i++){
        int x = build_frame[i][0];
        int y = build_frame[i][1];
        int a = build_frame[i][2];
        int b = build_frame[i][3];
        if(b == 1){
            if(a == 0){
                if(buildGi(y,x)){
                    board[y][x][a] = true;
                }
            }
            else if(a == 1){
                if(buildBo(y,x)){
                    board[y][x][a] = true;
                }
            }
        }
        else if (b == 0){          
            canRemove(y,x,a);
        }

    }


    for(int i = 0 ;i<=n;i++){
        for(int j =0;j<=n;j++){
            for(int k =0;k<2;k++){
            vector<int> v;
            if(board[j][i][k]){
                v.push_back(i);
                v.push_back(j);
                v.push_back(k);
            }

            if(v.size()>0) answer.push_back(v);
            }
        }
    }

    return answer;
}